src/libraries/System.Numerics.Tensors/src/System/Numerics/Tensors/netcore/TensorShape.cs

```diff
@@ -0,0 +1,56 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+
+namespace System.Numerics.Tensors
+{
+    internal readonly struct TensorShape
+    {
+        internal readonly nint[]? _metadata;      // 8 bytes
+
+        internal readonly nint _memoryLength;   // 8 bytes
```

Comment:
Given the layout and size concerns of this change, does the new type require an explicit `StructLayoutAttribute` annotation? Are we ok with use the default layout?

Reply:
> This struct contains a managed field (`nint[]? _metadata`) and as such is not blittable and will never have `StructLayout` respected, the runtime will treat it as `Auto` regardless (you can technically mark it as `LayoutKind.Explicit` and that would be respected, but that's a de-optimization and would cause many other issues).

Does Mono behave the same? I think it might respect Sequential for managed types.

Reply:
> Does Mono behave the same? I think it might respect Sequential for managed types.

That doesn't really matter. RyuJIT doesn't and so code cannot depend on the managed side being sequential, that is part of why we must use `InlineArray` (rather than declaring n-sequential fields).

This is likewise an internal struct, so it can't be used by developers in interop scenarios. Due to it being a struct containing a managed field, it likewise couldn't be used in interop without marshalling and so there is zero benefit to changing the layout from its default.

Reply:
> > Does Mono behave the same? I think it might respect Sequential for managed types.
> 
> That doesn't really matter. RyuJIT doesn't and so code cannot depend on the managed side being sequential, that is part of why we must use `InlineArray` (rather than declaring n-sequential fields).
> 
> This is likewise an internal struct, so it can't be used by developers in interop scenarios. Due to it being a struct containing a managed field, it likewise couldn't be used in interop without marshalling and so there is zero benefit to changing the layout from its default.

Yeah but I've meant that Auto could maybe still be beneficial for Mono.

Reply:
Mono should likely consider adjusting their layout algorithm to match the RyuJIT algorithm in that scenario.

There are thousands of structs across the BCL and we are not going to go and annotate every single one of them to be explicitly `Auto` because they contain some managed field.